[pipelining] Add pipeline stage test

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126735
* __->__ #126721
* #126812

Test tracer's and manual's stage creation by using a basic schedule (GPipe).

(Migrated from https://github.com/pytorch/PiPPy/blob/main/test/test_pipeline_stage.py)

Test command:
```
$ python test_stage.py
```

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

kwen2501 committed:
[pipelining] Add pipeline stage test

[ghstack-poisoned]

wconstab reviewed (approved):
thanks!

wconstab commented:
@pytorchbot merge

H-Huang commented on test/distributed/pipelining/test_stage.py:
```diff
@@ -0,0 +1,200 @@
+torch.manual_seed(0)
+
+
+class StageTest(MultiProcContinousTest):
```
Just curious, what does `MultiProcContinousTest` do? How is it different than `MultiProcessTestCase`

H-Huang reviewed (approved):
Nice!

kwen2501 committed:
Update on "[pipelining] Add pipeline stage test"


Test tracer's and manual's stage creation by using a basic schedule (GPipe).

(Migrated from https://github.com/pytorch/PiPPy/blob/main/test/test_pipeline_stage.py)

Test command:
```
$ python test_stage.py
```

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge

clee2000 commented:
@pytorchbot revert -m "The test_public_bindings failure is real, you just got unlucky since it was also broken on trunk for a different reason" -c ignoredsignal

kwen2501 committed:
Update on "[pipelining] Add pipeline stage test"


Test tracer's and manual's stage creation by using a basic schedule (GPipe).

(Migrated from https://github.com/pytorch/PiPPy/blob/main/test/test_pipeline_stage.py)

Test command:
```
$ python test_stage.py
```

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

kwen2501 committed:
Update on "[pipelining] Add pipeline stage test"


Test tracer's and manual's stage creation by using a basic schedule (GPipe).

(Migrated from https://github.com/pytorch/PiPPy/blob/main/test/test_pipeline_stage.py)

Test command:
```
$ python test_stage.py
```

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

kwen2501 committed:
Update on "[pipelining] Add pipeline stage test"


Test tracer's and manual's stage creation by using a basic schedule (GPipe).

(Migrated from https://github.com/pytorch/PiPPy/blob/main/test/test_pipeline_stage.py)

Test command:
```
$ python test_stage.py
```

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge
